src/piadm.sh

```diff
@@ -0,0 +1,741 @@
+    else
+	# Guaranteed that s0 is EFI System Partition, ZFS lives on s1.
+	suffix=s1
+    fi
+
+    some=0
+    for a in "${boot_devices[@]}"; do
+	# Plow through devices, even if some fail.  installboot also
+	# does loader-into-EFI-System-Partition this way.
+	# Trailing / is important in the -b argument because boot is actually
+	# a symlink.
+	installboot -m -b /${bootfs}/boot/ /${bootfs}/boot/pmbr \
+	    /${bootfs}/boot/gptzfsboot \
+	    /dev/rdsk/${a}${suffix} > /dev/null 2>&1 || \
+	    echo "WARNING: Can't installboot on ${a}${suffix}"
+	if [[ $? -eq 0 ]]; then
+	    some=1
+	fi
+    done
+    if [[ $some -eq 0 ]]; then
+	fatal "Could not installboot(1M) on ANY vdevs of pool $2"
+    fi
+
+}
+
+activate() {
+    pistamp=$1
+    piname_present_get_bootfs $pistamp $2
+    pool=$(echo $bootfs | awk -F/ '{print $1}')
+
+    cd /$bootfs
+    bootstamp=$(file -h platform | awk '{print $5}' | sed 's/\.\/platform-//g')
+    if [[ -d platform-$pistamp ]]; then
+	if [[ $bootstamp == $pistamp ]]; then
+	    vecho "NOTE: $pistamp is the current active PI."
+	    return
+	fi
+    else
+	echo "$pistamp is not a stamp for a PI on pool $pool"
+	usage
+    fi
+
+    vecho "Platform Image $pistamp will be loaded on next boot,"
+
+    # Okay, at this point we have the platform sorted out.  Let's see
+    # if we can do the same with the boot.
+    if [[ -d boot-$pistamp ]]; then
+	rm -f boot
+	ln -s ./boot-$pistamp boot
+	echo $pistamp > etc/version/boot
+	update_boot_sectors $pool $bootfs
+	grep -q 'fstype="ufs"' ./boot/loader.conf
+	if [[ $? -ne 0 ]]; then
+	    # Fix the loader.conf for keep-the-ramdisk booting.
+	    echo 'fstype="ufs"' >> ./boot/loader.conf
+	fi
+	vecho "    with a new boot image,"
+    else
+	vecho "    WARNING:  $pistamp has no matching boot image, using"
+    fi
+
+    vecho "    boot image " `cat etc/version/boot`
+
+    rm -f platform
+    ln -s ./platform-$pistamp platform
+}
+
+remove() {
+    pistamp=$1
+    piname_present_get_bootfs $pistamp $2
+    cd /$bootfs
+    bootstamp=$(file -h platform | awk '{print $5}' | sed 's/\.\/platform-//g')
+
+    if [[ -d platform-$pistamp ]]; then
+	if [[ $bootstamp == $pistamp ]]; then
+	    echo "$pistamp is the next-booting PI. Please activate another PI"
+	    echo "using 'piadm activate <other-PI-stamp>' first."
+	    usage
+	fi
+
+	# Boot image processing.
+	if [[ -d boot-$pistamp ]]; then
+	    # Boot bits may be older than the current PI, and the current PI
+	    # may not have matching boot bits for some reason.  Guard against
+	    # shooting yourself in the foot.
+	    grep -q $pistamp etc/version/boot
+	    if [[ $? -eq 0 ]]; then
+		# Oh no, pistamp points to the current boot bits.
+		echo "$pistamp is the current set of boot binaries.  Please"
+		echo "activate another pi using 'piadm activate <other-PI-stamp>' first."
+		usage
+	    fi
+	    /bin/rm -rf boot-$pistamp
+	fi
+
+	/bin/rm -rf platform-$pistamp
+    else
+	echo "$pistamp is not a stamp for a PI on pool $pool"
```

Comment:
generally, errors should be going to stderr

Reply:
`eecho` to the rescue.